### PR TITLE
Port automotive to SystemScalarConverter

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6790)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@soonho-tri for feature review, please.

**Comment by soonho-tri** [comment]
:lgtm: Tested on :apple:  as well.

BTW, do you have a plan to handle `SimpleCarToEulerFloatingJoint` separately (it has `DoToAutoDiffXd`).

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-:-Kr2Pk_Z-ASkHI24w_F7:bgz4rkr)*

**Comment by jwnimmer-tri** [comment]
> BTW, do you have a plan to handle SimpleCarToEulerFloatingJoint ...

Burninate.  See #6787.

**Comment by soonho-tri** [comment]
> Burninate. See #6787.

👍

**Comment by sherm1** [comment]
Platform :lgtm:, though PTAL at the comments below re `final`. Could these be non-final, perhaps with a note that extensions won't support scalar type conversion?

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/bicycle_car.h, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-Kr7BvjyDvgriDD3TCrV:-Kr7BvjyDvgriDD3TCrW:bklrj7f) ([raw file](https://github.com/robotlocomotion/drake/blob/71cf89c65e0972ad1933fc2869f3ee633ba6cb76/drake/automotive/bicycle_car.h#L65)):*
> ```objc
> /// @ingroup automotive_plants
> template <typename T>
> class BicycleCar final : public systems::LeafSystem<T> {
> ```

BTW must this be final? That prevents even the most innocuous extensions and lowers the overall utility of the class.

---

*[drake/automotive/simple_car.h, line 52 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-Kr7CLx1cWN_kHNOT6VF:-Kr7CLx1cWN_kHNOT6VG:bmntrqf) ([raw file](https://github.com/robotlocomotion/drake/blob/71cf89c65e0972ad1933fc2869f3ee633ba6cb76/drake/automotive/simple_car.h#L52)):*
> ```objc
> /// @ingroup automotive_plants
> template <typename T>
> class SimpleCar final : public systems::LeafSystem<T> {
> ```

BTW final makes me :cry:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-:-Kr7DAn1ywaBfI6TAA3F:b-ejuump)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/bicycle_car.h, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-Kr7BvjyDvgriDD3TCrV:-Kr7EZInOoGxv4yAy_5C:bkgpzpk) ([raw file](https://github.com/robotlocomotion/drake/blob/71cf89c65e0972ad1933fc2869f3ee633ba6cb76/drake/automotive/bicycle_car.h#L65)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW must this be final? That prevents even the most innocuous extensions and lowers the overall utility of the class.
</blockquote></details>

Per #6785, transmogrification _must always_ produce the exact type.  So if someone were to subclass this, their `ToAutoDiff` and `ToSymbolic` methods would throw an exception, instead of returning `nullptr`.  This is especially relevant because by default, sparsity will use symbolic form when it is available (as determined by whether or not it returns `nullptr`).  So, a simple`"HasDirectFeedthrough` call would throw an exception anytime a subclass of this were added to a `Diagram`.

We could remove the final and say "Good luck, YMMV" but I'd rather have the code be intrinsically correct.  If we find people wanting to subclass systems such as this one (which I would find _very odd_), we can at that point remove the `final` and make the transmogrification support optional.

So in short: I think `final` is a best practice for specific leaf systems.  If we need to let users change their behavior, that should be through things like Parameters, delegation, or composition -- _not_ by virtual overrides.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6790)*

**Comment by sherm1** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/bicycle_car.h, line 65 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6790#-Kr7BvjyDvgriDD3TCrV:-Kr7kcNBZS7kkGCoP3Bz:b2i7d9z) ([raw file](https://github.com/robotlocomotion/drake/blob/71cf89c65e0972ad1933fc2869f3ee633ba6cb76/drake/automotive/bicycle_car.h#L65)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Per #6785, transmogrification _must always_ produce the exact type.  So if someone were to subclass this, their `ToAutoDiff` and `ToSymbolic` methods would throw an exception, instead of returning `nullptr`.  This is especially relevant because by default, sparsity will use symbolic form when it is available (as determined by whether or not it returns `nullptr`).  So, a simple`"HasDirectFeedthrough` call would throw an exception anytime a subclass of this were added to a `Diagram`.

We could remove the final and say "Good luck, YMMV" but I'd rather have the code be intrinsically correct.  If we find people wanting to subclass systems such as this one (which I would find _very odd_), we can at that point remove the `final` and make the transmogrification support optional.

So in short: I think `final` is a best practice for specific leaf systems.  If we need to let users change their behavior, that should be through things like Parameters, delegation, or composition -- _not_ by virtual overrides.
</blockquote></details>

OK, thanks. I see your point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6790)*

**Comment by jwnimmer-tri** [comment]
Yeah, I have to go back and either fix (`final`) or enhance (allow disable) the transmog on some of the earlier porting PRs.  I hadn't figured out the subclassing hazard until #6743 made me write some proofs.
